pkg/scd/store/store.go

```diff
@@ -12,146 +11,99 @@ import (
+// data store.
 type Store interface {
 	OperationStore
 	SubscriptionStore
 }
 
-var (
-	_ Store = &DummyStore{}
-)
+type Transaction interface {
+	// Retrieve Store that operates within this Transaction.
+	Store() (Store, error)
 
-// DummyStore implements Store interface entirely with error-free no-ops
-type DummyStore struct {
-}
-
-func makeDummyOperation(id scdmodels.ID) *scdmodels.Operation {
-	var (
-		altLo = float32(11235)
-		altHi = float32(81321)
-		start = time.Now()
-		end   = start.Add(2 * time.Minute)
-	)
-
-	return &scdmodels.Operation{
-		ID:            id,
-		Version:       314,
-		OVN:           scdmodels.NewOVNFromTime(time.Now(), id.String()),
-		USSBaseURL:    "https://exampleuss.com/utm",
-		AltitudeLower: &altLo,
-		AltitudeUpper: &altHi,
-		StartTime:     &start,
-		EndTime:       &end,
-		State:         scdmodels.OperationStateAccepted,
-	}
+	// Commit commits all the operations performed on the Transactor so far.
+	Commit() error
 
+	// Rollback rolls back all the operations performed on the Transactor so far.
+	Rollback() error
 }
 
-// makeDummySubscription returns a dummy subscription instance with ID id.
-func makeDummySubscription(id scdmodels.ID) *scdmodels.Subscription {
-	altLo := float32(11235)
-	altHi := float32(81321)
-	result := &scdmodels.Subscription{
-		ID:                   id,
-		Version:              314,
-		NotificationIndex:    123,
-		BaseURL:              "https://exampleuss.com/utm",
-		AltitudeLo:           &altLo,
-		AltitudeHi:           &altHi,
-		NotifyForOperations:  true,
-		NotifyForConstraints: false,
-		ImplicitSubscription: true,
-		DependentOperations: []scdmodels.ID{
-			scdmodels.ID("c09bcff5-35a4-41de-9220-6c140a9857ee"),
-			scdmodels.ID("2cff1c62-cf1a-41ad-826b-d12dad432f21"),
-		},
-	}
-	return result
+type Transactor interface {
+	// Transact begins an atomic transaction
+	Transact(ctx context.Context) (Transaction, error)
 }
 
-// SearchSubscriptions is a stubbed implementation of SearchSubscriptions.
-func (s *DummyStore) SearchSubscriptions(ctx context.Context, cells s2.CellUnion, owner dssmodels.Owner) ([]*scdmodels.Subscription, error) {
-	subs := []*scdmodels.Subscription{
-		makeDummySubscription(scdmodels.ID("444eab15-8384-4e39-8589-5161689aee56")),
+// TransactionOperation is an application action involving one or more chained
+// Store actions joined by application logic.
+type TransactionOperation func(store Store) (err error)
+
+// PerformOperationWithRetries creates a Transaction from the Transactor,
+// attempts to perform the provided action, and retries this process again if
+// it fails in a retryable way.
+func PerformOperationWithRetries(ctx context.Context, transactor Transactor, operation TransactionOperation, retries int) error {
+	var err error
+	for i := 0; i <= retries; i++ {
+		// Prepare a Store for `operation` to act on
+		tx, err := transactor.Transact(ctx)
+		if err != nil {
+			return err
```

Comment:
All of the returns in this function are leaking transactions by not calling rollback

Reply:
Oops, thanks; fixed.